### ISSUE TITLE
How PyArmor does it

### ISSUE THREAD

**Issue by Make42** [dashingsoft/pyarmor#195]
There is a section https://pyarmor.readthedocs.io/en/latest/how-to-do.html how PyArmor works. But I am not quite sure about how it actually works and I would like to know. Let me get more specific what I would like to understand:

There are two files of interest as far as I understand: `dist/myscripty.py` (F1 in the following) and `dist/pytransform/_pytransform.so` (F2 in the following). F1 contains a byte -string and F1 contains some other code. As far as I understand F2 has been created by compiling my original file to Python bytecode and then doing [some stuff](https://pyarmor.readthedocs.io/en/latest/how-to-do.html#how-to-obfuscate-python-scripts) to it.

1) But is F2 then a Python bytecode file?
2) What is the byte-string in F1?
3) How do these work together?

Additional question: I have been thinking of using them together with [Nuitka](http://nuitka.net/) for protection, but if PyArmor works as I described above, I think this should not be possible. Is that right?

**Comment by jondy** [comment]
> But is F2 then a Python bytecode file?

No. It's binary file, built from c sources.

> What is the byte-string in F1?

First compile `.py` to `.pyc`, then encrypt `.pyc` by DES3, finally convert the encrypted bytes to python byte-string constant.

> How do these work together?

They are connected by he runtime module `pytransform/__init__.py`, refer to the definition of function `pyarmor_runtime` and `_load_library` in it

I'm not familiar with nuitka (pyarmor could pack the scripts with backend tool `PyInstaller). If the scripts have been obfuscated, any dependent module or package imported by the script could not be found by build tool (nuitak or PyInstaller). For PyInstaller, PyArmor does some extra work, I think nuitka need to do some extra work either, refer to
https://pyarmor.readthedocs.io/en/latest/how-to-do.html#how-to-pack-obfuscated-scripts

Also refer to #174

**Comment by Make42** [comment]
Ok, let me check if I got it right:

My `py`-file gets compiled into a `pyc`-file, then it gets encrypted, then it gets turned into a byte-string and put into F1. This is my code.

F2 is not related directly with my code, but this is your PyArmor code for decrypting my byte-string from F1. You wrote this in some language (maybe C?), compiled it into binary executable code and provide it here as the file `_pytransform.so`.

In order to make this whole work, you have the `__init__.py`-file. This file makes sure that F2 gets the byte-string from F1 correctly and decrypts it for execution according to the rules from the `license.so`-file and the `pytransform.key`-file.

Questions:
1. Did I get this right?
2. Here at github, I did not see any C code. Is the implementation of F2 closed source? (Which would be fine by me - I am just asking for the sake of understanding.)

(I will look into the Nuitka issue after lunch.)
